cdap-common/src/main/java/co/cask/cdap/common/security/AuthEnforceRewriter.java

```diff
@@ -0,0 +1,555 @@
+      generatorAdapter.loadThis();
+      // the setter method has only one argument which is what we want the field to be set to
+      generatorAdapter.loadArg(0);
+      generatorAdapter.putField(classType, name, type);
+      generatorAdapter.returnValue();
+      generatorAdapter.endMethod();
+    }
+
+    @Override
+    public void visitEnd() {
+      if (!methodAnnotations.isEmpty()) {
+        // If this class had method annotation then we need to generate the authenticaionContext and
+        // authorizationEnforcer field and their setters
+        generateFieldAndSetter(authorizationEnforcerFieldName, AUTHORIZATION_ENFORCER_TYPE);
+        generateFieldAndSetter(authenticationContextFieldName, AUTHENTICATION_CONTEXT_TYPE);
+      }
+      super.visitEnd();
+    }
+  }
+
+  /**
+   * Process {@link AnnotationNode} information collected on a method parameters
+   *
+   * @param parameterAnnotationNode a {@link Map} of parameter position and annotation details of that parameter
+   * @return a new {@link Map} where the key is the parameter name given in the annotation and value is the
+   * position of the parameter.
+   */
+  private static Map<String, Integer> processParameterAnnotationNode(Map<Integer, AnnotationNode>
+                                                                       parameterAnnotationNode) {
+    Map<String, Integer> parameterAnnotation = new HashMap<>();
+    for (Map.Entry<Integer, AnnotationNode> entry : parameterAnnotationNode.entrySet()) {
+      // the AnnotationNode of parameter will be an array of size 2 where 0 is "value" and 1 is the name provided in
+      // the annotation.
+      String paraName = (String) entry.getValue().values.get(1);
+      // We expect all parameters to have unique names to ensure that here
+      Preconditions.checkArgument(!parameterAnnotation.containsKey(paraName),
+                                  String.format("A parameter with name %s was already found at position %s." +
+                                                  " Please use unique names.", paraName,
+                                                parameterAnnotation.get(paraName)));
+      // store the parameter position with the unique name given to it
+      parameterAnnotation.put(paraName, entry.getKey());
+    }
+    return parameterAnnotation;
+  }
+
+  /**
+   * A class which can process {@link AuthEnforce} {@link AnnotationNode}
+   */
+  private static class AuthEnforceAnnotationNodeProcessor {
+
+    final Set<Action> actions = new HashSet<>();
+    Type enforceOn = null;
+    List<String> entities = null;
+
+    AuthEnforceAnnotationNodeProcessor(AnnotationNode annotationNode) {
+      List values = annotationNode.values;
+      // AuthEnforce AnnotationNode will have 6 values: 3 field names and another 3 field containing their values.
+      Preconditions.checkArgument(values.size() == 6, String.format("AuthEnforce annotation has three field and " +
```

Comment:
Preconditions support string formatting already, no need to use `String.format`. Simply do

```java
Preconditions.checkArgument(...., "Need three fields, but found %s", values.size());
```